File: MF2TT2MF/TrackBuilder.cpp
/**
 Copyright (c) 2018, Chanwoo Noh
 All rights reserved.
 */

#include "TrackBuilder.hpp"
namespace YKS {
    
std::shared_ptr<TrackBuilder> TrackBuilder::putEvent(std::shared_ptr<TE::TrackEvent> event)  {
    _eventList.push_back(event);
    return shared_from_this();
}

std::shared_ptr<TrackBuilder> TrackBuilder::putEvent(std::vector<std::shared_ptr<TE::TrackEvent>> eventList) {
    _eventList.insert(_eventList.end(), eventList.begin(), eventList.end());
    return shared_from_this();
}

std::vector<std::string> TrackBuilder::build() {
    std::vector<std::string> track;
    track.push_back("MTrk");
    auto deltaTime = 0;
    for (const auto &event : _eventList) {
        deltaTime = event->leadTime();
        auto line = std::to_string(deltaTime) + " " + event->value();
        track.push_back(line);
    }
    track.push_back("TrkEnd");
    return track;
}
    
} // namespace YKS;

File: rust/src/track_builder.rs
use crate::track_event::*;

pub struct TrackBuilder {
    channel: u8,
    events: Vec<Box<dyn TrackEvent>>,
}

impl TrackBuilder {
    pub fn new(channel: u8) -> Self {
        TrackBuilder {
            channel,
            events: Vec::new(),
        }
    }

    pub fn put_event(&mut self, event: Box<dyn TrackEvent>) -> &mut Self {
        self.events.push(event);
        self
    }

    pub fn put_events(&mut self, events: Vec<Box<dyn TrackEvent>>) -> &mut Self {
        self.events.extend(events);
        self
    }

    pub fn build(&self) -> Vec<String> {
        let mut result = Vec::new();
        result.push(format!("MTrk"));
        
        for event in &self.events {
            result.push(format!("{} {}", event.lead_time(), event.value()));
        }
        
        result
    }

    pub fn event_list(&self) -> &[Box<dyn TrackEvent>] {
        &self.events
    }
}